api/src/Rbac/Access/Gate.php

```
@@ -0,0 +1,237 @@
+     * @return GateContract
+     */
+    public function policy($class, $policy)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Determine if the given ability should be granted.
+     *
+     * @param  string $itemName
+     * @param  array|mixed $arguments
+     * @return bool
+     */
+    public function check($itemName, $arguments = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return true;
+        }
+
+        try {
+            //default roles check
+            if (! empty($this->defaultRoles) && $this->checkAccessRecursive($itemName, $arguments)) {
+                return true;
+            }
+
+            $user = $this->resolveUser();
+        } catch (UserNotFoundException $e) {
+            return false;
+        }
+
+        $assignments = $this->getStorage()->getAssignments($user->getAuthIdentifier());
+
+        return $this->checkAccessRecursive($itemName, $arguments, $assignments);
+    }
+
+    /**
+     * Get Gate service for particular user.
+     *
+     * @param Authenticatable $user
+     * @return static
+     */
+    public function forUser(Authenticatable $user)
+    {
+        return new static($this->getStorage(), function () use ($user) {return $user;}, $this->defaultRoles);
+    }
+
+    /**
+     * Check permissions and roles recursively.
+     *
+     * @param $itemName
+     * @param $params
+     * @param array $assignments
+     * @param array $itemStack
+     * @return bool
+     */
+    protected function checkAccessRecursive($itemName, $params, $assignments = [], $itemStack = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return false;
+        }
+
+        $itemStack[$item->name] = $item;
+
+        if (isset($assignments[$itemName]) || in_array($itemName, $this->defaultRoles)) {
+            foreach ($itemStack as $itemName => $item) {
+                if (! $this->executeRule($item, $params)) {
+                    return false;
```

Does this mean the rules are fail early? So the object must fulfil _all_ requirements?